grades/api.py

```diff
@@ -214,17 +214,22 @@ def generate_program_certificate(user, program):
         user (User): a Django user.
         program (programs.models.Program): program where the user is enrolled.
     """
-    mmtrack = get_mmtrack(user, program)
 
     if MicromastersProgramCertificate.objects.filter(user=user, program=program).exists():
         log.error('User [%s] already has a certificate for program [%s]', user, program)
         return
 
-    for course in program.course_set.all():
-        best_grade = mmtrack.get_best_final_grade_for_course(course)
+    courses_in_program_ids = set(program.course_set.all().values_list('id', flat=True))
+
+    courses_with_cert = set(
+        MicromastersCourseCertificate.objects.filter(
+            user=user,
+            course_id__in=courses_in_program_ids
+        ).values_list('course__id', flat=True))
+
+    if courses_in_program_ids.difference(courses_with_cert):
+        return
```

Comment:
Might be missing something, but I think this block could just be expressed like this:

```python
num_courses_with_certs = (
  MicromastersCourseCertificate.objects.filter(
    user=user,
    course__program=program
  )
  .values_list('course', flat=True)
  .distinct().count()
)
if num_courses_with_certs < program.course_set.count():
  return
```

Reply:
I think this is the same amount of db requests. But I like my version because it doesn't require the join of certificates, course, and program tables.

Reply:
Good point. I do still think the `set` logic isn't necessary. I think this would be a bit clearer

```python
program_course_ids = program.course_set.values_list('id', flat=True)
num_courses_with_cert = (
  MicromastersCourseCertificate.objects.filter(
    user=user,
    course_id__in=program_course_ids
  )
  .values_list('course__id', flat=True)
  .distinct().count()
)
if len(program_course_ids) > num_courses_with_cert:
  return
```

Reply:
If I don't use set() in the `program_course_ids`calculation the queryset will be evaluated twice, in getting `num_courses_with_cert` and later on `len(program_course_ids)`

Reply:
unless `program_course_ids` qset will get cached in the first evaluation....

Reply:
Yeah I think you're right, and even if the first query is cached, it's probably better to make it explicit. Either `list` or `set` for the first query will work then.